cosmoz-grouped-list.js

```diff
@@ -593,6 +590,82 @@
 			if (physicalIndex >= 0) {
 				return this._templateSelectors[physicalIndex].currentElement.__tmplInstance;
 			}
-		}
+		},
+		/**
+		 * Selects/Deselects an item.
+		 * @param {Object} item - The item.
+		 * @param {Boolean} select - True if the item should get selected.
+		 * @returns {undefined}
+		 */
+		_changeItemSelection(item, select) {
+			const model = this._getModelFromItem(item);
+			const setModelState = () => model && (model['selected'] = select);
+
+			if (select && !this.isItemSelected(item)) {
+				this.push('selectedItems', item);
+				setModelState();
+				return;
+			}
+
+			if (select) {
+				return;
+			}
+
+			const idx = this.selectedItems.indexOf(item);
+			if (idx < 0) {
+				return;
+			}
+			this.splice('selectedItems', idx, 1);
+			setModelState();
+		},
+		/**
+		 * Returns the group of an item.
+		 * @param {Object} item - The item.
+		 * @returns {Object} - The group of an item.
+		 */
+		_getGroupOfItem(item) {
+			const groupState = this._groupsMap;
+			if (!groupState) {
+				return;
+			}
+			return this.data.find(group => group.items.find(i => i === item));
+		},
+		/**
+		 * Checks if all items of a group are selected.
+		 * @param {Object} group - The group to make the check on.
+		 * @returns {Boolean} - True if all items of a group are selected.
+		 */
+		_allGroupItemsSelected(group) {
+			if (!group || !group.items) {
+				return;
+			}
+			return group.items.every(item => {
+				const model = this._getModelFromItem(item);
+				return model && model['selected'];
+			});
+		},
+
+		_selectedItemsChanged() {
+			this._debounceValidateGroupSelectionState();
```

Comment:
Why do we need to debounce here ? Is it because of the range select implementation in `cosmoz-omnitable` ?
In that case, it could be more efficient to provide some multi-select function in `cosmoz-group-list`, and validate group selection state right after that without debouncing.

Reply:
a) I think it is also valuable for non-range selections. If all items in a group get selected (one by one) the group should also get selected. 

b) Regarding debounce: Yes, you are right, because of the implementation in `cosmoz-omnitable`:
```
items.forEach(i => this.$.groupedList.selectItem(i));
```

Considering a & b I could call it without debouncing from `selectItem`, `deselectItem`,  and an additional method `selectItems` with the same outcome. Or as implemented with debounce on `selectedItems` observer.

What do you think @plequang ?

Reply:
Looking at how gmail multi-select is working (see my note https://github.com/Neovici/cosmoz-omnitable/issues/177#issuecomment-399955586), you could implement in cosmoz-group list a function like the following:

```
// toggle the selection of a range of items, from the specified start item to the specified end item
// the result selected state of this range should be the result of toggling the selection state of the end item
toggleSelectItems(fromItem, toItem) {
}
````
In `selectItem`, `deselectItem`, and `toggleSelectItems` you could call synchronously `_validateGroupSelectionState`. 

But if you think there is a benefit in observing `selectedItems` and deboucing the call to `_validateGroupSelectionState`, why not.